# Задание к лекции "Функции".
Для зачета необходимо правильно выполнить 4 заданий из 6.  

Для зачета с отличием – все 6.

## Задание 1

Дан список вида:

In [4]:
def _getsum(m):
    vl_sum = 0
    for x in range(0, len(m)):
        vl_sum += m[x][x]
    return vl_sum

In [5]:
data = [[13, 25, 23, 34],
        [45, 32, 44, 47],
        [12, 33, 23, 95],
        [13, 53, 34, 35],]

print(_getsum(data))

103


Напишите функцию, которая возвращает сумму элементов на диагонали. Т. е. 13+32+23+35.  
Список может быть любой длины, но всегда является "квадратным" (количество элементов во вложенных списках равно их количеству).

## Задание 2

Напишите функцию, возвращающую сумму первых n чисел Фибоначчи (для этого вам может понадобится ознакомиться с рекурсией).

In [6]:
def _fibosum(n):
    f_num = [0 ,1]
    v_tmp, v_1, v_2 = 0, 0, 1

    if n < 3:
        print('Числа Фибоначчи: ', f_num)
        return sum(f_num[:n])
    if n > 2:
        for a in range(3, n+1):
            v_tmp = v_2     #промежуточная переменная
            v_2 = v_1 + v_2 #значение следующего числа в ряду
            v_1 = v_tmp     #значение предыдущего числа в ряду
            f_num.append(v_2)
        print('Числа Фибоначчи: ', f_num)
        return sum(f_num[:n])

In [7]:
n = int(input('Введите количество первых n чисел Фибоначчи: '))
print('Сумма чисел =',_fibosum(n))

Введите количество первых n чисел Фибоначчи: 123
Числа Фибоначчи:  [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181, 6765, 10946, 17711, 28657, 46368, 75025, 121393, 196418, 317811, 514229, 832040, 1346269, 2178309, 3524578, 5702887, 9227465, 14930352, 24157817, 39088169, 63245986, 102334155, 165580141, 267914296, 433494437, 701408733, 1134903170, 1836311903, 2971215073, 4807526976, 7778742049, 12586269025, 20365011074, 32951280099, 53316291173, 86267571272, 139583862445, 225851433717, 365435296162, 591286729879, 956722026041, 1548008755920, 2504730781961, 4052739537881, 6557470319842, 10610209857723, 17167680177565, 27777890035288, 44945570212853, 72723460248141, 117669030460994, 190392490709135, 308061521170129, 498454011879264, 806515533049393, 1304969544928657, 2111485077978050, 3416454622906707, 5527939700884757, 8944394323791464, 14472334024676221, 23416728348467685, 37889062373143906, 61305790721611591, 99194853094755497, 160500643816367088, 25

## Задание 3

In [8]:
cook_book = {
  'салат': [
     {'ingridient_name': 'сыр', 'quantity': 50, 'measure': 'гр'},
     {'ingridient_name': 'томаты', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'огурцы', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'маслины', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'оливковое масло', 'quantity': 20, 'measure': 'мл'},
     {'ingridient_name': 'салат', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'перец', 'quantity': 20, 'measure': 'гр'}
    ],
  'пицца': [
     {'ingridient_name': 'сыр', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'колбаса', 'quantity': 30, 'measure': 'гр'},
     {'ingridient_name': 'бекон', 'quantity': 30, 'measure': 'гр'},
     {'ingridient_name': 'оливки', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'томаты', 'quantity': 20, 'measure': 'гр'},
     {'ingridient_name': 'тесто', 'quantity': 100, 'measure': 'гр'},   
    ],
  'лимонад': [
     {'ingridient_name': 'лимон', 'quantity': 1, 'measure': 'шт'},
     {'ingridient_name': 'вода', 'quantity': 200, 'measure': 'мл'},
     {'ingridient_name': 'сахар', 'quantity': 10, 'measure': 'гр'},
     {'ingridient_name': 'лайм', 'quantity': 20, 'measure': 'гр'},    
    ]
}

In [9]:
#продукты и их размерность
def _get_ingr_measure(order_list):
    ret_vl = {}
    for s in order_list:
        for c in cook_book[s]:
            ret_vl.setdefault(c['ingridient_name'], c['measure'])
    return ret_vl

#уникальные блюда и их количество в заказе
def _get_uniq_dish(order_list):
    ret_vl = {}
    for x in order_list:
        try:
            ret_vl[x] += 1
        except:
            ret_vl.setdefault(x, 1)
    return ret_vl

#количество ингридиентов
def _get_ingr_count(uniq_dish):
    ret_vl = {}
    for k, v in uniq_dish.items():
        for c in cook_book[k]:
            try:
                ret_vl[c['ingridient_name']] += c['quantity'] * v * guest_cnt
            except:
                ret_vl.setdefault(c['ingridient_name'], c['quantity'] * v * guest_cnt)
    return ret_vl

In [10]:
order_vl = input('Введите блюда через пробел: ').split(' ')
guest_cnt = int(input('Введите количество гостей: '))

ing_m = _get_ingr_measure(order_vl)
b = _get_uniq_dish(order_vl)
f = _get_ingr_count(b)

print('')
print('Список покупок:')
for k,v in f.items():
    print(k, v, ing_m[k])

Введите блюда через пробел: пицца пицца лимонад
Введите количество гостей: 3

Список покупок:
сыр 120 гр
колбаса 180 гр
бекон 180 гр
оливки 60 гр
томаты 120 гр
тесто 600 гр
лимон 3 шт
вода 600 мл
сахар 30 гр
лайм 60 гр


Напишите программу (она может состоять из нескольких функций), которая будет запрашивать у пользователя количество гостей и список блюд, которые будут подаваться каждому гостю. В результате программа должна выдавать список покупок, состоящий из ингредиентов, которые нужно купить в магазине (если ингредиенты повторяются, то их нужно записать одной строкой, а не отдельно).

Пример:

вызов\_главной\_функций()

Введите количество гостей: 3  
Введите блюда через пробел: пицца пицца лимонад

Список покупок:  
сыр 120 гр  
колбаса 180 гр  
бекон 180 гр  
оливки 60 гр  
томаты 120 гр  
тесто 600 гр  
лимон 3 шт  
вода 600 мл  
сахар 30 гр  
лайм 60 гр  

## Задание 4

Вы работаете секретарем и вам постоянно приходят различные документы. Вы должны быть очень внимательны, чтобы не потерять ни один документ. Каталог документов хранится в следующем виде:

In [11]:
documents = [
    {'type': 'passport', 'number': '2207 876234', 'name': 'Василий Гупкин'},
    {'type': 'invoice', 'number': '11-2', 'name': 'Геннадий Покемонов'},
    {'type': 'insurance', 'number': '10006', 'name': 'Аристарх Павлов'}
]

Перечень полок, на которых находятся документы хранится в следующем виде:

In [12]:
directories = {
    '1': ['2207 876234', '11-2', '5455 028765'],
    '2': ['10006', '5400 028765', '5455 002299'],
    '3': []
}

Необходимо реализовать пользовательские команды (внимание! это не название функций, которые должны быть выразительными, а команды, которые вводит пользователь, чтобы получить необходимый результат):
  
- p – people – команда, которая спросит номер документа и выведет имя человека, которому он принадлежит;  
- s – shelf – команда, которая спросит номер документа и выведет номер полки, на которой он находится  
Правильно обработайте ситуацию, когда пользователь будет вводить несуществующий документ.

## Задание 5

К условию из задания 5 реализуйте следующие команды:
  
- l – list – команда, которая выведет список всех документов в формате   
**passport "2207 876234" "Василий Гупкин"**  
- as – add shelf – команда, которая спросит номер новой полки и добавит ее в перечень

## Задание 6

К условию из задания 5 реализуйте следующие команды:  
  
- d – delete – команда, которая спросит номер документа и удалит его из каталога и из перечня полок. Предусмотрите сценарий, когда пользователь вводит несуществующий документ;  
- m – move – команда, которая спросит номер документа и целевую полку и переместит его с текущей полки на целевую. Корректно обработайте кейсы, когда пользователь пытается переместить несуществующий документ или переместить документ на несуществующую полку;  
- a – add – команда, которая добавит новый документ в каталог и в перечень полок, спросив его номер, тип, имя владельца и номер полки, на котором он будет храниться.

In [1]:
documents = [{'type': 'passport', 'number': '2207 876234', 'name': 'Василий Гупкин'},
             {'type': 'invoice', 'number': '11-2', 'name': 'Геннадий Покемонов'},
             {'type': 'insurance', 'number': '10006', 'name': 'Аристарх Павлов'}]
directories = {'1': ['2207 876234', '11-2', '5455 028765'],
               '2': ['10006', '5400 028765', '5455 002299'],
               '3': []}

#запрос номера документа
def _get_doc_num():
    #ret_vl = input('Введите номер документа: ')
    ret_vl = '2207 876234'
    return ret_vl

#запрос номера полки
def _get_sh_num():
    #ret_vl = input('Введите номер полки: ')
    ret_vl = '3'
    return ret_vl

#запрос данных
def _get_vl(q_str):
    #ret_vl = input(q_str)
    #return ret_vl
    return input(q_str)

#проверка наличия документа
def _check_doc(doc_num):
    ret_vl = False
    for d in documents:
        if doc_num == d['number']:
            ret_vl = True
    return ret_vl

#проверка наличия полки
def _check_st(s_num):
    ret_vl = False
    for d in directories.keys():
        if s_num == d:
            ret_vl = True
    return ret_vl

#p - поиск человека
def _get_people():
    doc_num = _get_vl('Укажите номер документа: ')
    
    ret_vl = 'Введен несуществующий документ'
    for d in documents:
        if doc_num == d['number']:
            ret_vl = d['name']
    print(ret_vl)

#s - поиск полки
def _get_directory():
    doc_num = _get_vl('Укажите номер документа: ')
    
    ret_vl = 'Введен несуществующий документ'
    for key, value in directories.items():
        if doc_num in value:
            ret_vl = key
    print(ret_vl)

#l - список документов
def _view_all_docs():
    for d in documents:
        print(d['type'], '\"' + d['number'] + '\"', '\"' + d['name'] + '\"')

#as - новый номер полки
def _add_new_sh():
    st_num = _get_vl('Укажите номер полки для добавления: ')
    
    if st_num not in directories.keys(): 
        directories[st_num] = []
    else:
        print("Ошибка добавления: полка с номером " + st_num + " уже существует")

#d - удаление записи
def _del_doc():
    doc_num = _get_vl('Укажите номер документа для удаления: ')
    
    if _check_doc(doc_num):
        for d in documents:
            if doc_num == d['number']:
                documents.remove(d)
                ret_vl = 'Запись для документа удалена'
                break

        for d in directories:
            if doc_num in directories[d]:
                directories[d].remove(doc_num)
                break
    else:
        ret_vl = 'Ошибка удаления: попытка удалить несуществующий документ'
    print(ret_vl)

#m - перенос
def _migr_doc():
    st_num = _get_vl('Укажите номер полки для хранения документа: ')
    doc_num = _get_vl('Укажите номер документа: ')
    
    msg_doc = _check_doc(doc_num)
    msg_s = _check_st(st_num)
    
    if not (msg_doc and msg_s):
        print ('ошибка')
    else:
        for d in directories:
            if doc_num in directories[d]:
                directories[d].remove(doc_num)
                directories[st_num].append(doc_num)

#a - добавление новой записи
def _add_new():
    st_num = _get_vl('Укажите номер полки для хранения документа: ')
    if not _check_st(st_num):
        print('Указана несуществующая полка')
        return
    
    doc_num = _get_vl('Укажите номер документа: ')
    if _check_doc(doc_num):
        print('Указан существующий номер документа')
        return
    
    doc_type = _get_vl('Укажите тип документа: ')
    nm = _get_vl('Укажите имя владельца: ')
    
    msg_doc = _check_doc(doc_num)
    msg_s = _check_st(st_num)
    
    documents.append({'type': doc_type, 'number': doc_num, 'name': nm})
    directories[st_num].append(doc_num)

In [ ]:
print('-'*20)
print('Команды: ')
print('p - поиск человека по номеру документа')
print('s - поиск номера полки по номеру документа')
print('l - вывести список всех документов')
print('as - добавить новый номер полки')
print('d - удалить документ из каталога и перечня полок')
print('m - перемещение документа на полку')
print('a - добавление нового документа')
print('-'*20)
print('')
a = True

while a:
    cmd = input('Введите команду: ')
    if cmd == 'p':
        _get_people()
    elif cmd == 's':
        _get_directory()
    elif cmd == 'l':
        _view_all_docs()
    elif cmd == 'as':
        _add_new_sh()
    elif cmd == 'd':
        _del_doc()
    elif cmd == 'm':
        _migr_doc()
    elif cmd == 'a':
        _add_new()
    else:
        print('введенная команда не существует')
    
    print('')
    if input('Продолжить работу? (y/n)') != 'y':
        a = False

--------------------
Команды: 
p - поиск человека по номеру документа
s - поиск номера полки по номеру документа
l - вывести список всех документов
as - добавить новый номер полки
d - удалить документ из каталога и перечня полок
m - перемещение документа на полку
a - добавление нового документа
--------------------

Введите команду: a
Укажите номер полки для хранения документа: 3
Укажите номер документа: 11-2
Указан существующий номер документа

Продолжить работу? (y/n)y
Введите команду: a
Укажите номер полки для хранения документа: 3
Укажите номер документа: 123
Укажите тип документа: oass
Укажите имя владельца: test top

Продолжить работу? (y/n)y
Введите команду: l
passport "2207 876234" "Василий Гупкин"
invoice "11-2" "Геннадий Покемонов"
insurance "10006" "Аристарх Павлов"
oass "123" "test top"

Продолжить работу? (y/n)y
Введите команду: d
Укажите номер документа для удаления: 123
Запись для документа удалена

Продолжить работу? (y/n)y
Введите команду: l
passport "2207 876234" "Вас